perllib/Open311/GetServiceRequests.pm

```diff
@@ -0,0 +1,193 @@
+        };
+        if ($@) {
+            warn "Not creating problem $request_id for @{[$body->name]}, bad update time"
+                if $self->verbose;
+            next;
+        }
+
+        my $updated = DateTime::Format::W3CDTF->format_datetime(
+            $updated_time->clone->set_time_zone('UTC')
+        );
+        if (@args && ($updated lt $args[0] || $updated gt $args[1]) ) {
+            warn "Problem id $request_id for @{[$body->name]} has an invalid time, not creating"
+                if $self->verbose;
+            next;
+        }
+
+        my $created_time = eval {
+            DateTime::Format::W3CDTF->parse_datetime(
+                $request->{requested_datetime} || ""
+            )->set_time_zone(
+                FixMyStreet->time_zone || FixMyStreet->local_time_zone
+            );
+        };
+        $created_time = $updated_time if $@;
+
+        my $problems;
+        my $criteria = {
+            external_id => $request_id,
+        };
+        $problems = $self->schema->resultset('Problem')->to_body($body)->search( $criteria );
+
+        my @contacts = grep { $request->{service_code} eq $_->category } $contacts->all;
+        my $contact = $contacts[0] ? $contacts[0]->category : 'Other';
+
+        my $state = $self->map_state($request->{status});
+
+        unless (my $p = $problems->first) {
+            my $problem = $self->schema->resultset('Problem')->new(
+                {
+                    user => $self->system_user,
+                    external_id => $request_id,
+                    detail => $request->{description},
+                    title => $request->{title} || $request->{service_name} . ' problem',
+                    anonymous => 0,
+                    name => $self->system_user->name,
+                    confirmed => $created_time,
+                    created => $created_time,
+                    lastupdate => $updated_time,
+                    whensent => $created_time,
+                    state => $state,
+                    postcode => '',
+                    used_map => 1,
+                    latitude => $request->{lat},
+                    longitude => $request->{long},
+                    areas => ',' . $body->id . ',',
+                    bodies_str => $body->id,
+                    send_method_used => 'Open311',
+                    category => $contact,
+                }
+            );
+
+            if ($request->{media_url}) {
+                my $ua = LWP::UserAgent->new;
+                my $res = $ua->get($request->{media_url});
+                if ( $res->is_success && $res->content_type eq 'image/jpeg' ) {
+                    my $photoset = FixMyStreet::App::Model::PhotoSet->new({
+                        data_items => [ $res->decoded_content ],
+                    });
+                    $problem->photo($photoset->data);
+                }
+            }
+
+            $problem->insert();
+        }
+    }
+
+    return 1;
+}
+
+sub map_state {
```

Comment:
This is a copy of the function in GetServiceRequestUpdates, so should be shared somewhere (in Open311.pm?).